File: src/main/java/com/bazbatlabs/smashballs/models/World.java
package com.bazbatlabs.smashballs.models;

import java.util.*;

public final class World {

    public static final float WIDTH = 400.0f;
    public static final float HEIGHT = 300.0f;

    private static final int INITIAL_CHANCES = 30;
    private static final int BRICKS_PER_ROW = 16;
    private static final float BRICK_WIDTH = WIDTH / BRICKS_PER_ROW;
    private static final float BRICK_HEIGHT = BRICK_WIDTH / 3.0f;

    private final Rect bounds;
    private final Paddle paddle;
    private final Ball ball;
    private final List<Wall> walls;
    private final List<Brick> bricks;
    private final WorldEvents events;
    private State state;
    private int chancesRemaining;

    public World(WorldEvents events) {
        Vec2 origin = Vec2.ZERO;
        Vec2 size = new Vec2(WIDTH, HEIGHT);

        this.bounds = new Rect(origin, size);
        this.events = events;

        this.chancesRemaining = INITIAL_CHANCES;

        this.walls = new ArrayList<Wall>();

        this.walls.add(new Wall(new Rect(origin, new Vec2(0.0f, size.y)), events));
        this.walls.add(new Wall(new Rect(new Vec2(origin.x, origin.y + size.y),
                                         new Vec2(size.x, 0.0f)), events));
        this.walls.add(new Wall(new Rect(new Vec2(origin.x + size.x, origin.y),
                                         new Vec2(0.0f, size.y)), events));

        Vec2 brickSize = new Vec2(BRICK_WIDTH, BRICK_HEIGHT);

        this.bricks = new ArrayList<Brick>();
        for (int i = 0; i < BRICKS_PER_ROW; i++) {
            Vec2 pos = new Vec2(origin.x + (i * BRICK_WIDTH), 250.0f);
            this.bricks.add(new Brick(new Rect(pos, brickSize), Brick.Type.TOUGH, events));
        }

        this.paddle = new Paddle(bounds, 50f, events);
        this.ball = new Ball(paddle.center(), this);
        this.state = State.NORMAL;
    }

    public float floor() { return bounds.bottom(); }
    public Rect bounds() { return bounds; }

    public Paddle paddle() { return paddle; }
    public Ball ball() { return ball; }
    public List<Brick> bricks() { return bricks; }

    public void update() {
        if (state == State.NORMAL) {
            paddle.update();
            ball.update();

            for (Brick brick : bricks) {
                brick.update();
            }

            clearDestroyedBricks();

            if (bricks.isEmpty()) {
                state = State.CLEARED;
            }

            if (ball.isLost()) {

                chancesRemaining--;

                if (chancesRemaining > 0) {
                    ball.reset(paddle.center());
                } else {
                    state = State.GAME_OVER;
                }
            }
        }
    }

    private void clearDestroyedBricks() {
        ListIterator<Brick> iter = bricks.listIterator();
        while (iter.hasNext()) {
            if (iter.next().isDestroyed()) {
                iter.remove();
            }
        }
    }

    public List<Collidable> collidables() {
        List<Collidable> collidables =  new ArrayList<Collidable>();

        collidables.addAll(walls);
        collidables.addAll(bricks);
        collidables.add(paddle);

        return collidables;
    }

    public void startAcceleratingPaddle(Direction direction) {
        paddle.startAccelerating(direction);
    }

    public void stopAcceleratingPaddle(Direction direction) {
        paddle.stopAccelerating(direction);
    }

    public void kickstartBall() {
        Vec2 vel = paddle.vel();
        float angle = 0.0f;

        if (vel.x > 0.0f) {
            angle = (float)Math.PI / 4f;
        } else if (vel.x < 0.0f) {
            angle = 3f * (float)Math.PI / 4f;
        } else {
            angle = (float)Math.PI / 2f;
        }

        ball.kickstart(angle);
    }

    public boolean isGameOver() { return state == State.GAME_OVER; }
    public boolean isCleared() { return state == State.CLEARED; }

    private enum State {
        NORMAL, CLEARED, GAME_OVER
    }
}

File: src/main/scala/models/World.scala
package com.bazbatlabs.bricksmash.models

import com.bazbatlabs.bricksmash.lib._

class World(brickLayout: Array[Array[Option[Brick.Type.Value]]], val events: WorldEvents) {

  val StateChangeDuration = 100

  val InitialChances = 30
  val BricksPerRow = 14

  val PaddleStartHeight = World.Height / 9f
  val BrickStartHeight = 25f * World.Height / 28f

  val BrickWidth = World.Width / BricksPerRow
  val BrickHeight = BrickWidth / 2f

  val origin = Vec2.Zero
  val size = Vec2(World.Width, World.Height)

  val bounds = Rect(origin, size)

  var chancesRemaining = InitialChances

  val paddle = new Paddle(bounds, PaddleStartHeight, events)
  val ball = new Ball(paddle.center, this)

  var state = State.Normal
  var stateCounter = 0

  val walls = List(new Wall(Rect(origin, Vec2(0f, size.y)), events),
                   new Wall(Rect(Vec2(origin.x, origin.y + size.y),
                                 Vec2(size.x, 0f)), events),
                   new Wall(Rect(Vec2(origin.x + size.x, origin.y),
                                 Vec2(0f, size.y)), events))

  val brickSize = Vec2(BrickWidth, BrickHeight)

  var bricks = List[Brick]()

  var y = BrickStartHeight
  for (line <- brickLayout) {
    var x = 0f

    for (brickType <- line) {
      if (!brickType.isEmpty) {
        val pos = Vec2(origin.x + x, y)
        bricks = new Brick(Rect(pos, brickSize), brickType.get, events) :: bricks
      }

      x += BrickWidth
    }

    y -= BrickHeight
  }

  def floor = bounds.bottom

  def update() {
    if (state == State.Normal) {
      paddle.update()
      ball.update()

      for (brick <- bricks) {
        brick.update()
      }

      clearDestroyedBricks

      if (bricks.isEmpty) {
        state = State.Cleared
        stateCounter = 0
      }

      if (ball.isLost) {
        chancesRemaining -= 1

        if (chancesRemaining > 0) {
          ball.reset(paddle.center)
        } else {
          state = State.GameOver
          stateCounter = 0
        }
      }
    }

    stateCounter += 1
  }

  def clearDestroyedBricks {
    bricks = bricks.filterNot { brick => brick.isDestroyed }
  }

  def collidables = paddle :: walls ::: bricks

  def startAcceleratingPaddle(dir: Direction.Value) {
    paddle.startAccelerating(dir)
  }

  def stopAcceleratingPaddle(dir: Direction.Value) {
    paddle.stopAccelerating(dir)
  }

  def isReadyToMoveToNextLevel: Boolean =
    state == State.Cleared && stateCounter > StateChangeDuration

  def isReadyToResetToMenu: Boolean =
    state == State.GameOver && stateCounter > StateChangeDuration

  def kickstartBall() {
    val pi = Math.PI.asInstanceOf[Float]
    val angle = if (paddle.vel.x > 0f) pi / 4f
                else if (paddle.vel.x < 0f) 3f * pi / 4f
                else pi / 2f

    ball.kickstart(angle)
  }

  def isGameOver = state == State.GameOver
  def isCleared = state == State.Cleared


  object State extends Enumeration {
    type State = Value
    val Normal = Value("Normal")
    val Cleared = Value("Cleared")
    val GameOver = Value("GameOver")
  }
}

object World {
  val Width = 224.0f
  val Height = 186.0f
}
